File: scripts/generate-icons.js
#!/usr/bin/env node

/**
 * Generate PWA icons for Converty
 *
 * This script generates all required icon sizes for Progressive Web App support:
 * - 192x192px: Chrome minimum requirement
 * - 512x512px: Chrome minimum requirement
 * - 192x192px maskable: Android adaptive icon with 80% safe zone
 * - 180x180px: iOS home screen (apple-touch-icon)
 *
 * Current implementation creates a simple branded icon programmatically.
 * Replace the generateSourceIcon() function with actual branded assets later.
 */

const sharp = require("sharp");
const fs = require("node:fs");
const path = require("node:path");

// Output directory
const ICONS_DIR = path.join(__dirname, "..", "public", "icons");

/**
 * Generate a simple source icon (1024x1024) for Converty
 * This is a placeholder that creates a gradient background with "C" letter
 * Replace this with actual branded SVG/PNG assets when available
 */
async function generateSourceIcon() {
  const size = 1024;
  const centerX = size / 2;
  const centerY = size / 2;

  // Create SVG with gradient background and "C" letter
  const svg = `
    <svg width="${size}" height="${size}" xmlns="http://www.w3.org/2000/svg">
      <defs>
        <linearGradient id="grad" x1="0%" y1="0%" x2="100%" y2="100%">
          <stop offset="0%" style="stop-color:#4F46E5;stop-opacity:1" />
          <stop offset="100%" style="stop-color:#7C3AED;stop-opacity:1" />
        </linearGradient>
      </defs>
      <rect width="${size}" height="${size}" fill="url(#grad)" rx="180"/>
      <text
        x="${centerX}"
        y="${centerY + 120}"
        font-family="Arial, sans-serif"
        font-size="600"
        font-weight="bold"
        fill="white"
        text-anchor="middle">C</text>
    </svg>
  `;

  return Buffer.from(svg);
}

/**
 * Generate maskable icon with 80% safe zone
 * Maskable icons need padding to ensure content stays within the safe zone
 * when adaptive icons apply various shapes (circle, rounded square, squircle)
 */
async function generateMaskableIcon(sourceBuffer, outputSize) {
  const safeZoneRatio = 0.8; // 80% safe zone as per maskable icon spec
  const contentSize = Math.round(outputSize * safeZoneRatio);
  const padding = Math.round((outputSize - contentSize) / 2);

  // Create solid background
  const background = await sharp({
    create: {
      width: outputSize,
      height: outputSize,
      channels: 4,
      background: { r: 79, g: 70, b: 229, alpha: 1 }, // Match gradient start color
    },
  })
    .png()
    .toBuffer();

  // Resize source to fit in safe zone
  const resizedContent = await sharp(sourceBuffer)
    .resize(contentSize, contentSize, {
      fit: "contain",
      background: { r: 0, g: 0, b: 0, alpha: 0 },
    })
    .toBuffer();

  // Composite resized content centered on background
  return sharp(background)
    .composite([
      {
        input: resizedContent,
        top: padding,
        left: padding,
      },
    ])
    .png()
    .toBuffer();
}

/**
 * Main icon generation function
 */
async function generateIcons() {
  console.log("🎨 Generating PWA icons for Converty...\n");

  // Create icons directory if it doesn't exist
  if (!fs.existsSync(ICONS_DIR)) {
    fs.mkdirSync(ICONS_DIR, { recursive: true });
    console.log("✓ Created icons directory:", ICONS_DIR);
  }

  // Generate source icon
  console.log("Generating source icon...");
  const sourceIcon = await generateSourceIcon();

  // Icon configurations
  const icons = [
    { name: "icon-192x192.png", size: 192, maskable: false },
    { name: "icon-512x512.png", size: 512, maskable: false },
    { name: "icon-192-maskable.png", size: 192, maskable: true },
    { name: "apple-touch-icon.png", size: 180, maskable: false },
  ];

  // Generate each icon
  for (const icon of icons) {
    const outputPath = path.join(ICONS_DIR, icon.name);

    let buffer;
    if (icon.maskable) {
      buffer = await generateMaskableIcon(sourceIcon, icon.size);
    } else {
      buffer = await sharp(sourceIcon)
        .resize(icon.size, icon.size, {
          fit: "contain",
          background: { r: 0, g: 0, b: 0, alpha: 0 },
        })
        .png()
        .toBuffer();
    }

    await fs.promises.writeFile(outputPath, buffer);
    console.log(
      `✓ Generated ${icon.name} (${icon.size}x${icon.size}${icon.maskable ? ", maskable" : ""})`
    );
  }

  console.log("\n✅ All PWA icons generated successfully!");
  console.log(`📁 Icons location: ${ICONS_DIR}`);
  console.log(
    "\nℹ️  These are placeholder icons. Replace generateSourceIcon() with branded assets when available."
  );
}

// Run the script
generateIcons().catch((err) => {
  console.error("❌ Error generating icons:", err);
  process.exit(1);
});

File: scripts/generate-icons.py
import base64
import os
from io import BytesIO
from PIL import Image

SRC = r"C:\Users\admin\Desktop\666-e743b96e-edf0-42f4-abef-f0c33117b367.jpg"
PUBLIC = r"C:\Users\admin\Desktop\kaifa\converty-site\public"


def center_crop_square(img: Image.Image) -> Image.Image:
    w, h = img.size
    size = min(w, h)
    left = (w - size) // 2
    top = (h - size) // 2
    return img.crop((left, top, left + size, top + size))


def to_png_data_uri(img: Image.Image) -> str:
    buf = BytesIO()
    img.save(buf, format="PNG")
    b64 = base64.b64encode(buf.getvalue()).decode("ascii")
    return f"data:image/png;base64,{b64}"


def make_svg(data_uri: str, width: int, height: int) -> str:
    return (
        f'<svg xmlns="http://www.w3.org/2000/svg" width="{width}" height="{height}" '
        f'viewBox="0 0 {width} {height}">'
        f'<image href="{data_uri}" width="{width}" height="{height}" />'
        f'</svg>'
    )


def main():
    img = Image.open(SRC).convert("RGBA")
    square = center_crop_square(img)

    os.makedirs(os.path.join(PUBLIC, "icons"), exist_ok=True)

    # 256 logo
    logo = square.resize((256, 256), Image.Resampling.LANCZOS)
    logo_path = os.path.join(PUBLIC, "logo.png")
    logo.save(logo_path, format="PNG")
    print(f"saved {logo_path}")

    # favicon.svg / logo.svg (base64 embedded PNG for sharp rendering at any size)
    data_uri = to_png_data_uri(logo)
    for name, size in [("favicon.svg", 32), ("logo.svg", 256)]:
        svg_path = os.path.join(PUBLIC, name)
        with open(svg_path, "w", encoding="utf-8") as f:
            f.write(make_svg(data_uri, size, size))
        print(f"saved {svg_path}")

    # favicon.ico multi-size
    ico_sizes = [16, 32, 48, 64]
    ico_imgs = [square.resize((s, s), Image.Resampling.LANCZOS) for s in ico_sizes]
    ico_path = os.path.join(PUBLIC, "favicon.ico")
    ico_imgs[0].save(
        ico_path,
        format="ICO",
        sizes=[(s, s) for s in ico_sizes],
        append_images=ico_imgs[1:],
    )
    print(f"saved {ico_path}")

    # PWA icons
    icon_192 = square.resize((192, 192), Image.Resampling.LANCZOS)
    icon_192.save(os.path.join(PUBLIC, "icons", "icon-192x192.png"), format="PNG")
    print("saved icons/icon-192x192.png")

    icon_512 = square.resize((512, 512), Image.Resampling.LANCZOS)
    icon_512.save(os.path.join(PUBLIC, "icons", "icon-512x512.png"), format="PNG")
    print("saved icons/icon-512x512.png")

    # Apple touch icon
    icon_180 = square.resize((180, 180), Image.Resampling.LANCZOS)
    icon_180.save(os.path.join(PUBLIC, "icons", "apple-touch-icon.png"), format="PNG")
    print("saved icons/apple-touch-icon.png")

    # Maskable icon: keep content within center 80% safe zone, use 70% content to be safe
    mask_size = 192
    padding = int(mask_size * 0.15)
    content_size = mask_size - 2 * padding
    maskable = Image.new("RGBA", (mask_size, mask_size), (0, 0, 0, 0))
    content = square.resize((content_size, content_size), Image.Resampling.LANCZOS)
    maskable.paste(content, (padding, padding), content)
    maskable.save(os.path.join(PUBLIC, "icons", "icon-192-maskable.png"), format="PNG")
    print("saved icons/icon-192-maskable.png")


if __name__ == "__main__":
    main()
